packages/server/src/fhir/search.ts

```diff
@@ -113,6 +114,57 @@ export async function searchImpl<T extends Resource>(
   };
 }
 
+export async function searchByReferenceImpl<T extends Resource>(
+  repo: Repository,
+  searchRequest: SearchRequest<T>,
+  referenceField: string,
+  references: string[]
+): Promise<Record<string, T[]>> {
+  validateSearchResourceTypes(repo, searchRequest);
+
+  if (searchRequest.count === undefined) {
+    searchRequest.count = DEFAULT_SEARCH_COUNT;
+  } else if (searchRequest.count > maxSearchResults) {
+    searchRequest.count = maxSearchResults;
+  }
+
+  const results: Record<string, T[]> = {};
+  const resourceType = searchRequest.resourceType;
+  const builder = new SelectPerReferenceQuery(
+    searchRequest.resourceType,
+    new Column(resourceType, referenceField),
+    references
+  );
+  builder
+    .column({ tableName: resourceType, columnName: 'id' })
+    .column({ tableName: resourceType, columnName: 'content' });
+  repo.addDeletedFilter(builder as unknown as SelectQuery);
+  repo.addSecurityFilters(builder as unknown as SelectQuery, resourceType);
```

Comment:
Given that the `SelectPerReferenceQuery` actually has several `SelectQuery` internally, we'd need to make sure that we're propagating all of these security filters to all of the inner queries.

I haven't parsed that fully, maybe you're already handling those?

```ts
    this.referenceQuery = new SelectQuery(this.tableName);
    this.resultsInnerQuery = new SelectQuery(this.tableName);
    this.resultsQuery = new SelectQuery('ranked', this.resultsInnerQuery);
    this.outerQuery = new SelectQuery('references', this.referenceQuery);
```

Reply:
The security filters are propagated correctly. To do this the "right" way, it would be ideal to have a targeted interface that both `SelectQuery` and `SelectPerReferenceQuery` implements. Currently, I believe the interface would just need to specify the `where` method.